packages/ragbits-document-search/src/ragbits/document_search/ingestion/document_processor.py

```diff
@@ -7,30 +8,30 @@
 from ragbits.document_search.ingestion.providers.unstructured.images import UnstructuredImageProvider
 from ragbits.document_search.ingestion.providers.unstructured.pdf import UnstructuredPdfProvider
 
-ProvidersConfig = dict[DocumentType, BaseProvider]
+ProvidersConfig = dict[DocumentType, type[BaseProvider] | BaseProvider]
```

Comment:
i'm not sure if we want to have distinct value type here, however - assuming later comment that we would potentially like to cache instantiated providers we would need both

Reply:
Suggestion(non-blocking): maybe instead of typing this as `type[BaseProvider] | BaseProvider` we could type it as `Callable[[], LLM] | BaseProvider`? This would allow us to accept all three: provider objects, provider classes, factory functions that return provider objects. And if I'm not mistaken, this could be made without doing any changes to the code logic, just the typing - the code itself seems to be already ready for factory functions.

Reply:
I think having an union of types here is okay.

On the other hand, I'm always nervous when using unions with dicts because dicts being [invariant](https://blog.daftcode.pl/covariance-contravariance-and-invariance-the-ultimate-python-guide-8fabc0c24278) causes them to behave in ways that are not always intuitive when it comes to unions (for example, forcing you to use `cast` in line 69, which normally wouldn't be necessary). So maybe we can think about changing `dict` to some specialised data class - but I don't think it's a priority right now